### PR TITLE
[VarExporter] a new component to serialize values to plain PHP code

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

This PR proposes moving what is currently the `PhpMarshaller` class in the Cache component to a separate component.

This component would provide only one public static method:
`VarExporter::export($value, bool &$isStaticValue = null): string`.

This method returns `$value` serialized as plain PHP code. Running this code creates the same exact data structure that `$value` contained. This is exactly like `serialize()` and `unserialize()`, from which all semantics are preserved (`__sleep`, `__wakeup` and `Serializable`).

The reason to use this method *vs* `serialize()` or even igbinary is performance: thanks to OPcache, the resulting code is significantly faster and more memory efficient than using `unserialize()` or `igbinary_unserialize()`.

Unlike `var_export()`, this works on any serializable PHP value.

It also provides a few improvements over `var_export()`/`serialize()`:
- the output is PSR-2 compatible
- the output can be re-indented without messing up with any `\r` or `\n` in the data
- missing classes throw a `ReflectionException` instead of being unserialized to a `PHP_Incomplete_Class` object
- references involving `SplObjectStorage`, `ArrayObject` or `ArrayIterator` instances are preserved
- `Reflection*`, `IteratorIterator` and `RecursiveIteratorIterator` classes throw an exception when being serialized (their unserialized version is broken anyway, see  https://bugs.php.net/76737.)

### CODE REVIEW

**Comment by tolry** [comment]
since the naming is similar/identical it might be a good idea to distinguish this from the php core function http://php.net/manual/de/function.var-export.php ?

**Comment by nicolas-grekas** [comment]
@tolry we have the `VarDumper` component that provides a better `var_dump()` function. I think that the `VarExporter` name conveys what it should: this is a better `var_export()`.

**Comment by tolry** [comment]
@nicolas-grekas I guessed so, just thought it might be easier to grasp, if the explanation mentioned var_export explicitely. Ah, I see it's in there now - did you add it or did I miss it before? :grin:

**Comment by nicolas-grekas** [comment]
@tolry  yes, I updated the description.

PR is ready IMHO.

Here is what I added to the description above:
- the output is PSR-2 compatible
- the output can be re-indented without messing up with any `\r` or `\n` in the data
- missing classes throw a `ReflectionException` instead of being unserialized to a `PHP_Incomplete_Class` object
- references involving `SplObjectStorage`, `ArrayObject` or `ArrayIterator` instances are preserved
- `Reflection*`, `IteratorIterator` and `RecursiveIteratorIterator` classes throw an exception when being serialized (their unserialized version is broken anyway, see  https://bugs.php.net/76737.)

File: src/Symfony/Component/VarExporter/Internal/Exporter.php [hunk 1]
```diff
@@ -9,17 +9,17 @@
  * file that was distributed with this source code.
  */
 
-namespace Symfony\Component\Cache\Marshaller\PhpMarshaller;
+namespace Symfony\Component\VarExporter\Internal;
 
 /**
  * @author Nicolas Grekas <p@tchwork.com>
  *
  * @internal
```
Comment (Koc) [inline src/Symfony/Component/VarExporter/Internal/Exporter.php]: Does this class stay internal after moving it to separate component?

File: src/Symfony/Component/VarExporter/VarExporter.php [hunk 2]
```diff
@@ -9,40 +9,38 @@
...
-use Symfony\Component\Cache\Marshaller\PhpMarshaller\Registry;
-use Symfony\Component\Cache\Marshaller\PhpMarshaller\Values;
+use Symfony\Component\VarExporter\Internal\Configurator;
+use Symfony\Component\VarExporter\Internal\Exporter;
+use Symfony\Component\VarExporter\Internal\Reference;
+use Symfony\Component\VarExporter\Internal\Registry;
+use Symfony\Component\VarExporter\Internal\Values;
 
 /**
  * @author Nicolas Grekas <p@tchwork.com>
```
Comment (sroze) [inline src/Symfony/Component/VarExporter/VarExporter.php]: We usually put that at the end of the comment.

File: src/Symfony/Component/VarExporter/composer.json [hunk 3]
```diff
@@ -0,0 +1,36 @@
+{
+    "name": "symfony/var-exporter",
+    "type": "library",
+    "description": "A blend of var_export() + serialize() to turn any serializable data structure to plain PHP code",
+    "keywords": ["export", "serialize"],
```
Comment (sroze) [inline src/Symfony/Component/VarExporter/composer.json]: I'd add `symfony` and `var-exporter` as well, they are the component 💃

File: src/Symfony/Component/VarExporter/CHANGELOG.md [hunk 4]
```diff
@@ -0,0 +1,7 @@
+CHANGELOG
+=========
+
+4.2.0
+-----
+
+ * added the component
```
Comment (sroze) [inline src/Symfony/Component/VarExporter/CHANGELOG.md]:
```patch
-  * added the component
+ * Created the component (extracted `PhpMarshaller` from the `Cache` component)
```

Could be valuable to give a hint of where it comes from.

**Review by sroze — APPROVED** [review]
Except these two comments, it looks good to me 👌 

Note that it's very likely it will require more maintenance than when it was internal.

File: src/Symfony/Component/VarExporter/composer.json (hunk 3 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/VarExporter/composer.json]: this is already in the name of the component, no need to duplicate

File: src/Symfony/Component/VarExporter/CHANGELOG.md (hunk 4 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/VarExporter/CHANGELOG.md]: PhpMarshaller has never been released, better not give it any existence except in internal history
I'd keep as is :)

File: src/Symfony/Component/VarExporter/VarExporter.php (hunk 2 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/VarExporter/VarExporter.php]: moved down

**Comment by nicolas-grekas** [comment]
> Note that it's very likely it will require more maintenance than when it was internal.

That's why I added more tests for more cases ;)

Comments addressed, thanks for the review.

File: src/Symfony/Component/VarExporter/README.md [hunk 5]
```diff
@@ -0,0 +1,33 @@
+VarExporter Component
+=====================
+
+The VarExporter component allows exporting any serializable PHP data structure to
+plain PHP code. While doing so, it preserves all the semantics associated to the
```
Comment (fabpot) [inline src/Symfony/Component/VarExporter/README.md]: associated with

File: src/Symfony/Component/VarExporter/README.md [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+serialization mechanism of PHP (`__wakeup`, `__sleep`, `Serializable` esp.)
```
Comment (fabpot) [inline src/Symfony/Component/VarExporter/README.md]: "esp" can be removed IMHO

File: src/Symfony/Component/VarExporter/README.md [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+The reason to use this component *vs* `serialize()` or
+[igbinary](https://github.com/igbinary/igbinary) is performance: thanks to
+OPcache, the resulting code is significantly faster and more memory efficient
+than using `unserialize()` or `igbinary_unserialize()`.
+
+Unlike `var_export()`, this works on any serializable PHP value.
+
+It also provides a few improvements over `var_export()`/`serialize()`:
+ * the output is PSR-2 compatible
```
Comment (fabpot) [inline src/Symfony/Component/VarExporter/README.md]: missing empty line before the list, and missing `;` at the end of each line but the last.

File: src/Symfony/Component/VarExporter/README.md [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+ * the output can be re-indented without messing up with `\r` or `\n` in the data
+ * missing classes throw a `ReflectionException` instead of being unserialized to
+   `PHP_Incomplete_Class` objects
+ * references involving `SplObjectStorage`, `ArrayObject` or `ArrayIterator`
+   instances are preserved
+ * `Reflection*`, `IteratorIterator` and `RecursiveIteratorIterator` classes
+   throw an exception when being serialized (their unserialized version is broken
+   anyway, see  https://bugs.php.net/76737.)
```
Comment (fabpot) [inline src/Symfony/Component/VarExporter/README.md]: dot outside the parenthesis

Comment (fabpot) [reply src/Symfony/Component/VarExporter/README.md]: extra space after "see"

File: src/Symfony/Component/VarExporter/VarExporter.php [hunk 9]
```diff
@@ -9,40 +9,50 @@
...
+use Symfony\Component\VarExporter\Internal\Values;
 
 /**
- * @author Nicolas Grekas <p@tchwork.com>
+ * Exports serializable PHP values to PHP code.
  *
- * PhpMarshaller allows serializing PHP data structures using var_export()
- * while preserving all the semantics associated to serialize().
+ * VarExporter allows serializing PHP data structures to plain PHP code (like var_export())
+ * while preserving all the semantics associated to serialize() (unlike var_export()).
```
Comment (fabpot) [inline src/Symfony/Component/VarExporter/VarExporter.php]: associated with

File: src/Symfony/Component/VarExporter/VarExporter.php [hunk 10]
```diff
@@ -9,40 +9,50 @@
...
-    public static function marshall($value, bool &$isStaticValue = null): string
+    /**
+     * Exports a serializable PHP value to PHP code.
+     *
+     * @param mixed $value          The value to export
+     * @param bool  &$isStaticValue Set to true after execution if the provided value is static, false otherwise
+     *
+     * @return string The value exported as PHP code
+     *
+     * @throw \Exception When the provided value cannot be serialized
```
Comment (fabpot) [inline src/Symfony/Component/VarExporter/VarExporter.php]: @throws

**Review by fabpot — CHANGES_REQUESTED** [review]

**Comment by nicolas-grekas** [comment]
Comments addressed thanks.

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.

**Pull request merged**